src/main/java/org/beanpod/switchboard/dao/UserDaoImpl.java

```diff
@@ -0,0 +1,40 @@
+package org.beanpod.switchboard.dao;
+
+import java.text.MessageFormat;
+import java.util.Optional;
+import javax.validation.ValidationException;
+import lombok.RequiredArgsConstructor;
+import org.beanpod.switchboard.dto.UserDto;
+import org.beanpod.switchboard.dto.mapper.UserMapper;
+import org.beanpod.switchboard.entity.SwitchBoardUserDetails;
+import org.beanpod.switchboard.entity.UserEntity;
+import org.beanpod.switchboard.repository.UserRepository;
+import org.openapitools.model.UserModel;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class UserDaoImpl {
+
+  private final UserRepository userRepository;
+  private final UserMapper userMapper;
+
+  public UserDto save(UserModel userModel) {
+    if (userRepository.findByUsername(userModel.getUsername()).isPresent()) {
+      throw new ValidationException("Username exists!");
+    }
+    return userMapper.toUserDto(userRepository.save(userMapper.toUserEntity(userModel)));
+  }
+
+  public UserDetails loadUserByUsername(String username) {
+    final Optional<UserEntity> optionalUser = userRepository.findByUsername(username);
+    if (optionalUser.isPresent()) {
+      return new SwitchBoardUserDetails(optionalUser.get());
+    } else {
+      throw new UsernameNotFoundException(
+          MessageFormat.format("User with username {0} cannot be found.", username));
```

Comment:
Using a more generic message such as "bad credentials" will get rid of one of your security hotspots.

Reply:
sonar still says its a security issue after fixing, it actually didnt use to in the more previous commits although i had the same old logic. I dont know if im missing something

Reply:
[resolved](https://github.com/bean-pod/switchboard/pull/305#pullrequestreview-590786489)